File: nodes/filter-messages/filter-messages.process.py
from typing import Any, Dict, List, Union

async def process(inputs: Dict[str, Any], settings: Dict[str, Any], config: Dict[str, Any], nodeConfig: Dict[str, Any]) -> Dict[str, Any]:
    try:
        messages = inputs.get('messages')
        filter_type = inputs.get('filter_type')
        filter_value = inputs.get('filter_value')

        if not isinstance(messages, list):
            raise Exception("Messages input must be an array")

        if not isinstance(filter_type, str) or not filter_type.strip():
            raise Exception("Filter type must be a non-empty string")

        if filter_value is None:
            raise Exception("Filter value is required")

        valid_filter_types = [
            'min_length', 'max_length', 'contains_text', 
            'starts_with', 'ends_with', 'role_equals', 'role_not_equals'
        ]

        if filter_type not in valid_filter_types:
            raise Exception(f"Invalid filter type: {filter_type}. Must be one of: {', '.join(valid_filter_types)}")

        # Helper function to extract text content from message
        def get_message_text(message):
            if not isinstance(message, dict):
                return ''
            content = message.get('content', '')
            if isinstance(content, str):
                return content
            elif isinstance(content, list):
                text_parts = []
                for item in content:
                    if isinstance(item, dict) and item.get('type') == 'text' and 'text' in item:
                        text_parts.append(item['text'])
                return ' '.join(text_parts)
            else:
                return str(content)

        # Filter messages based on type
        filtered_messages = []
        for message in messages:
            if not isinstance(message, dict):
                continue

            if filter_type == 'min_length':
                min_length = int(filter_value) if isinstance(filter_value, (int, float)) else 0
                if get_message_text(message).__len__() >= min_length:
                    filtered_messages.append(message)

            elif filter_type == 'max_length':
                max_length = int(filter_value) if isinstance(filter_value, (int, float)) else 0
                if get_message_text(message).__len__() <= max_length:
                    filtered_messages.append(message)

            elif filter_type == 'contains_text':
                contains_text = str(filter_value).lower()
                if contains_text in get_message_text(message).lower():
                    filtered_messages.append(message)

            elif filter_type == 'starts_with':
                starts_with = str(filter_value).lower()
                if get_message_text(message).lower().startswith(starts_with):
                    filtered_messages.append(message)

            elif filter_type == 'ends_with':
                ends_with = str(filter_value).lower()
                if get_message_text(message).lower().endswith(ends_with):
                    filtered_messages.append(message)

            elif filter_type == 'role_equals':
                if message.get('role') == filter_value:
                    filtered_messages.append(message)

            elif filter_type == 'role_not_equals':
                if message.get('role') != filter_value:
                    filtered_messages.append(message)

        return {
            "messages": filtered_messages,
            "count": len(filtered_messages)
        }

    except Exception as error:
        print(f'Filter Messages error: {error}')
        raise Exception(f"Filter Messages error: {str(error)}")

File: sdks/nodejs/nodes/filter-messages/filter-messages.process.js
export default async ({inputs, settings, config}) => {
    try {
        const messages = inputs.messages;
        const filterType = inputs.filter_type;
        const filterValue = inputs.filter_value;

        if (!Array.isArray(messages)) {
            throw new Error("Messages input must be an array");
        }

        if (typeof filterType !== 'string' || filterType.trim() === '') {
            throw new Error("Filter type must be a non-empty string");
        }

        if (filterValue === undefined || filterValue === null) {
            throw new Error("Filter value is required");
        }

        const validFilterTypes = [
            'min_length', 'max_length', 'contains_text', 
            'starts_with', 'ends_with', 'role_equals', 'role_not_equals'
        ];

        if (!validFilterTypes.includes(filterType)) {
            throw new Error(`Invalid filter type: ${filterType}. Must be one of: ${validFilterTypes.join(', ')}`);
        }

        // Helper function to extract text content from message
        const getMessageText = (message) => {
            if (!message || typeof message !== 'object') return '';
            if (typeof message.content === 'string') return message.content;
            if (Array.isArray(message.content)) {
                return message.content
                    .filter(item => item && item.type === 'text' && item.text)
                    .map(item => item.text)
                    .join(' ');
            }
            return String(message.content || '');
        };

        // Filter messages based on type
        const filteredMessages = messages.filter(message => {
            if (!message || typeof message !== 'object') return false;

            switch (filterType) {
                case 'min_length':
                    const minLength = Number(filterValue);
                    if (isNaN(minLength)) return false;
                    return getMessageText(message).length >= minLength;

                case 'max_length':
                    const maxLength = Number(filterValue);
                    if (isNaN(maxLength)) return false;
                    return getMessageText(message).length <= maxLength;

                case 'contains_text':
                    const containsText = String(filterValue).toLowerCase();
                    return getMessageText(message).toLowerCase().includes(containsText);

                case 'starts_with':
                    const startsWith = String(filterValue).toLowerCase();
                    return getMessageText(message).toLowerCase().startsWith(startsWith);

                case 'ends_with':
                    const endsWith = String(filterValue).toLowerCase();
                    return getMessageText(message).toLowerCase().endsWith(endsWith);

                case 'role_equals':
                    return message.role === filterValue;

                case 'role_not_equals':
                    return message.role !== filterValue;

                default:
                    return false;
            }
        });

        return {
            messages: filteredMessages,
            count: filteredMessages.length
        };

    } catch (error) {
        console.log('Filter Messages error:', error);
        throw new Error(`Filter Messages error: ${error.message}`);
    }
};
